tests/test_response.py

```diff
@@ -1288,3 +1288,26 @@ def test_set_content_type():
     res = Response(content_type='application/json')
     res.content_type = 'application/foo'
     assert res.content_type == 'application/foo'
+
+def test_raises_no_charset():
+    with pytest.raises(TypeError):
+        Response(content_type='image/jpeg', body=text_(b'test'))
+
+def test_raises_none_charset():
+    with pytest.raises(TypeError):
+        Response(
+            content_type='image/jpeg',
+            body=text_(b'test'),
+            charset=None)
+
+def test_doesnt_raise_with_charset_content_type_has_no_charset():
+    res = Response(content_type='image/jpeg', body=text_(b'test'), charset='utf-8')
+    assert res.body == b'test'
+    assert res.content_type == 'image/jpeg'
+    assert res.charset is None
+
+def test_conten_type_has_charset():
+    res = Response(content_type='application/foo; charset=UTF-8', body=text_(b'test'))
+    assert res.body == b'test'
+    assert res.content_type == 'application/foo'
+    assert res.charset == 'UTF-8'
```

Comment:
Yes. `r.content_type` can be set with a full `content_type` string and everything works correctly. As in `r.headers['Content-Type'] == 'application/foo; charset=UTF-8'`. `r.content_type` only returns the content-type not the content-type with parameters.

Certain properties have some really bad assymetric API's. Unfortunately they have been that way for a long time, can't change that now.